README.md
#
# NLIB Perl Application Framework
#

This is a very hand-crafted application framework purely written in Perl.

Features:

- various application types, such as CLI, web or API application
- session management
- templating
- ORM-like entity representation

Feel free to make use of it!


ChangeLog
2014-03-10 21:30  nils-1327

	* api/htdocs/nlib.js:
	  switcher from GET to POST XHR requests

2014-03-07 22:36  nils-1327

	* api/htdocs/nlib.js:
	  re-added cyclic stringify object handling

2014-03-07 14:01  nils-1327

	* api/htdocs/nlib.js:
	  changes

2014-02-28 12:40  nils-1327

	* api/htdocs/nlib.js:
	  modifications

2014-02-28 09:23  nils-1327

	* api/htdocs/nlib.js:
	  mods

2014-02-28 09:18  nils-1327

	* api/: NLIB/TableObject.pm, htdocs/nlib.js:
	  mods

2014-02-27 17:18  nils-1327

	* api/: NLIB/DbConnector.pm, NLIB/Session.pm, NLIB/SubScope.pm,
	  htdocs/nlib.js:
	  diverse changes

2014-02-18 14:25  nils-1327

	* api/htdocs/nlib.js:
	  addded updateView and rebind

2014-02-14 15:32  nils-1327

	* api/NLIB/TableObjectUser.pm:
	  Made table property overridable

2014-02-05 13:33  nils-1327

	* TODO, api/NLIB/Session.pm, api/NLIB/TableObject.pm:
	  Mods

2014-02-05 00:05  nils-1327

	* api/NLIB/TableObjectRequest.pm:
	  Mods

2014-02-04 10:28  nils-1327

	* api/NLIB/Application.pm, scripts/rollout.sh:
	  Implemented child session creation in constructor

2014-02-04 09:42  nils-1327

	* api/NLIB/Constants.pm:
	  Extended exception types

2014-02-04 00:17  nils-1327

	* Makefile, api/NLIB/Session.pm, api/NLIB/SubScope.pm:
	  Implemented session login handling and authentication update

2014-02-03 22:30  nils-1327

	* Makefile, sbin/reqgen/reqgen.csv:
	  Some mods

2014-02-03 22:30  nils-1327

	* api/htdocs/nlib.js:
	  Added debug functionality

2014-02-03 14:11  nils-1327

	* api/htdocs/nlib.js:
	  Fixed IE8 XML HTTP handling

2014-01-31 19:01  nils-1327

	* Makefile, TODO, api/NLIB/Action.pm, api/NLIB/ActionDispatcher.pm,
	  api/NLIB/ActionLogout.pm, api/NLIB/ActionWidgetState.pm,
	  api/NLIB/Application.pm, api/NLIB/Constants.pm,
	  api/NLIB/Content.pm, api/NLIB/ContentLogin.pm,
	  api/NLIB/ContentLogout.pm, api/NLIB/ContentWidgetState.pm,
	  api/NLIB/DbConnector.pm, api/NLIB/Dispatcher.pm,
	  api/NLIB/Exception.pm, api/NLIB/Flow.pm,
	  api/NLIB/FlowAuthenticate.pm, api/NLIB/FlowDispatcher.pm,
	  api/NLIB/JsonPage.pm, api/NLIB/Session.pm,
	  api/NLIB/TableObject.pm, api/NLIB/TableObjectRequest.pm,
	  api/NLIB/TableObjectUser.pm, api/NLIB/Template.pm,
	  api/tpl/nlibMainTemplate.htm, bin/nlib.cgi,
	  doc/nlib-kommunikationsstruktur.lyx,
	  doc/flowdefs/flow_anonymous.dot,
	  sbin/reqgen/generate_requests.pl, sbin/reqgen/reqgen.csv,
	  scripts/rollout.sh, sql/nlib-001.sql:
	  mods

2014-01-31 18:55  nils-1327

	* api/htdocs/nlib-app.js:
	  added

2014-01-31 18:55  nils-1327

	* api/htdocs/nlib.js:
	  Added ie8 Event handler

2014-01-30 11:22  nils-1327

	* doc/flowdefs/parse_flow.pl, sbin/reqgen/generate_requests.pl,
	  sbin/reqgen/reqgen.csv, sbin/reqgen/reqgen.pl:
	  added

2014-01-22 13:38  nils-1327

	* api/NLIB/: ContentLogin.pm, ContentLogout.pm:
	  added

2014-01-22 13:37  nils-1327

	* api/NLIB/ActionLogin.pm, api/NLIB/ActionLogout.pm,
	  api/NLIB/Flow.pm, api/NLIB/FlowAuthenticate.pm,
	  doc/application.dot, doc/flowdefs/parse_flow.pl,
	  sql/nlib-001.sql:
	  Example FlowAuthenticate implementation

2014-01-18 01:30  nils-1327

	* api/NLIB/ActionDispatcher.pm, api/NLIB/Dispatcher.pm,
	  api/NLIB/Flow.pm, api/NLIB/FlowAuthenticate.pm,
	  api/NLIB/FlowDispatcher.pm, api/NLIB/Session.pm,
	  api/NLIB/SubScope.pm, api/NLIB/TableObject.pm, bin/nlib.cgi:
	  Implemented first WebFlow

2014-01-17 14:41  nils-1327

	* api/NLIB/ActionDispatcher.pm, api/NLIB/FlowAuthenticate.pm,
	  bin/nlib.cgi:
	  Implementing first flow action (login)

2014-01-17 12:59  nils-1327

	* api/htdocs/htm/widgetLogin.htm:
	  added

2014-01-17 12:58  nils-1327

	* README, TODO, api/htdocs/nlib.js, api/tpl/nlibMainTemplate.htm,
	  scripts/rollout.sh:
	  Modifications

2014-01-15 14:02  nils-1327

	* api/htdocs/: nlib.js, htm/login.htm:
	  View handling tests

2014-01-15 09:22  nils-1327

	* api/: NLIB/JsonPage.pm, htdocs/nlib.js:
	  mods

2014-01-14 14:12  nils-1327

	* doc/nlib-kommunikationsstruktur.lyx:
	  added

2014-01-14 14:12  nils-1327

	* api/NLIB/Session.pm, api/htdocs/nlib.js, bin/nlib.cgi,
	  sql/nlib-001.sql:
	  fixed cookie jar handling. changed flow_status table. moved js
	  object methods to prototype property

2014-01-13 08:09  nils-1327

	* TODO, api/NLIB/Action.pm:
	  Cookie missing in JsonPage?

2014-01-09 14:04  nils-1327

	* api/NLIB/JsonPage.pm:
	  added

2014-01-09 14:03  nils-1327

	* api/NLIB.pm, api/NLIB/Action.pm, api/NLIB/Content.pm,
	  api/NLIB/Session.pm, api/htdocs/nlib.js, bin/nlib.cgi:
	  Separated between json and html content page

2014-01-08 21:10  nils-1327

	* api/htdocs/nlib.js, scripts/rollout.sh:
	  something

2014-01-08 13:43  nils-1327

	* api/htdocs/htm/login.htm, doc/application.dot:
	  added

2014-01-08 13:42  nils-1327

	* README, api/NLIB/Content.pm, api/NLIB/Dispatcher.pm,
	  api/htdocs/nlib.js, api/tpl/nlibMainTemplate.htm, bin/nlib.cgi:
	  Widget communication

2014-01-07 14:25  nils-1327

	* api/htdocs/: tempo.js, htm/ol.htm:
	  added

2014-01-07 14:24  nils-1327

	* api/htdocs/nlib.js, api/tpl/nlibException.htm,
	  api/tpl/nlibMainTemplate.htm, scripts/rollout.sh:
	  Added Tempo template handling

2014-01-06 22:56  nils-1327

	* api/: NLIB/Content.pm, htdocs/nlib.js, tpl/nlibMainTemplate.htm:
	  JS

2014-01-06 19:04  nils-1327

	* api/: NLIB/Content.pm, htdocs/nlib.js:
	  JS tests

2014-01-06 14:58  nils-1327

	* api/htdocs/nlib.js:
	  added

2014-01-06 14:58  nils-1327

	* TODO, api/tpl/nlibException.htm, api/tpl/nlibMainTemplate.htm:
	  JS tests

2014-01-05 23:20  nils-1327

	* api/NLIB/TableObjectRequest.pm, doc/flowdefs/flow.dot:
	  added

2014-01-05 23:19  nils-1327

	* TODO, api/NLIB.pm, api/NLIB/Action.pm,
	  api/NLIB/ActionDispatcher.pm, api/NLIB/Application.pm,
	  api/NLIB/Constants.pm, api/NLIB/Content.pm,
	  api/NLIB/Dispatcher.pm, api/NLIB/Flow.pm,
	  api/NLIB/FlowAuthenticate.pm, api/NLIB/Object.pm,
	  api/NLIB/Page.pm, bin/nlib.cgi, sql/nlib-001.sql:
	  Holiday modifications

2013-12-31 01:48  nils-1327


htdocs/nlib.js
/*
 * $Id: nlib.js,v 1.22 2014-03-10 20:30:57 nils-1327 Exp $ 
 */

var NLIB_DEBUG       = 0x0001;
var NLIB_INFO        = 0x0002;
var NLIB_ERROR       = 0x0004;
var NLIB_JSON        = 0x0008;

var NLIBDEBUG = NLIB_DEBUG | NLIB_INFO | NLIB_ERROR | NLIB_JSON;

NlibObject.prototype = new Object;
NlibObject.prototype.constructor = NlibObject;
function NlibObject(){
	this.xmlhttp = null;
  
	if (window.XMLHttpRequest) {
		this.xmlhttp = new window.XMLHttpRequest();
	}
	else{
		try{
			this.xmlhttp = new ActiveXObject("Microsoft.XMLHTTP.3.0");
		}
		catch(ex){
			this.debug(NLIB_CONSTRUCT,"new","Exception: "+ex);
		}
	}

	this.addReadyStateListener(this);
} // NlibObject

NlibObject.prototype.debug = function(level,meth,msg){
	var l = "";
	if(!console){return;}
	
	if((NLIBDEBUG & level) == NLIB_DEBUG )
		l = "[NLIB_DEBUG] ";

	else if((NLIBDEBUG & level) == NLIB_INFO )
		l = "[NLIB_INFO]  ";

	else if((NLIBDEBUG & level) == NLIB_ERROR )
		l = "[NLIB_ERROR] ";
	
	else if((NLIBDEBUG & level) == NLIB_JSON )
		l = "[NLIB_JSON] ";

	if(l != "")
	{
		if(meth && msg){ console.log(l+": "+this.constructor.name+"::"+meth+": "+msg); }
		else if(meth){ console.log(l+": "+this.constructor.name+"::"+meth); }
		else{ console.log(l+": "+this.constructor.name); }
	}


}

NlibObject.prototype.addReadyStateListener = function(obj){
		this.debug(NLIB_INFO,"addReadyStateListener");

		if(obj.xmlhttp)
		{
			if (!obj.xmlhttp.addEventListener)
			{
				obj.xmlhttp.onreadystatechange =  function(){
					if(obj.xmlhttp.readyState == 4 && obj.xmlhttp.status == 200 ){ obj.parseResponse(obj.xmlhttp.responseText); }
				};
			}
			else{
				obj.xmlhttp.addEventListener("readystatechange",function(){
					if(obj.xmlhttp.readyState == 4 && obj.xmlhttp.status == 200){ obj.parseResponse(obj.xmlhttp.responseText); }
				});
			}
		}
	}

/*
 * This method is for debugging JSON data.
 */
NlibObject.prototype.debugJSON = function(data){
	return JSON.stringify(data);
}


NlibObject.prototype.encodeRequest = function(data){
		this.debug(NLIB_INFO,"encodeRequest");
		$.base64.utf8encode = true;
		//var json = JSON;
		if(data)
		{
			//this.debug(NLIB_DEBUG,"encodeRequest","Returning "+$.base64('encode',json.stringify(data)));

			seen = []

			json = JSON.stringify(data, function(key, val) {
				if (typeof val == "object") {
					if (seen.indexOf(val) >= 0)
						return;
					seen.push(val)
				}
				return val;
			});

			//return $.base64('encode',JSON.stringify(data));
			return $.base64('encode',json);
		}
		else	
		{
			//this.debug(NLIB_DEBUG,"encodeRequest","Returning "+$.base64('encode',json.stringify(this.data)));
			return $.base64('encode',JSON.stringify(this.data));
		}
	}


NlibObject.prototype.sendRequest = function(req,val){
	this.debug(NLIB_INFO,"sendRequest");
	if(this.xmlhttp){
		var params = null;
		if(val)
			params = req+"="+val;
		//this.xmlhttp.open("GET", "%%%MAKEFILE_CGIBIN%%%?"+req+"="+val, true);
		
		this.xmlhttp.open("POST", "%%%MAKEFILE_CGIBIN%%%", true);
		this.xmlhttp.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
//		}
//		else
//			this.xmlhttp.open("GET", "%%%MAKEFILE_CGIBIN%%%?", true);
		this.xmlhttp.send(params);
	}

}
	
NlibObject.prototype.parseResponse = function(resp){
	this.debug(NLIB_INFO,"parseResponse");
};



/*
 * A common widget class. Inherit from this and override parseResponse
 * to handle the server response accordingly.
 */
NlibWidget.prototype = new NlibObject;	
NlibWidget.prototype.constructor = NlibWidget;
function NlibWidget(widgetName,app,data){
	this.debug(NLIB_INFO,"new");

	NlibObject.call(this);

	this.widgetName = widgetName;
	this.app = app;
	this.id_target = "#content"+this.widgetName; // id_target;
	this.id_widget = "widget"+this.widgetName; // id_widget;
	this.src = new String("%%%MAKEFILE_WIDGET_SOURCE%%%/widget"+this.widgetName+".htm");
	this.data = data;
	this.template = null;
}

NlibWidget.prototype.update = function(){
	this.debug(NLIB_INFO,"update");
	var data = new Object();
	data.widgetName = this.widgetName;
	this.sendRequest("updateWidget",this.encodeRequest(data));
}

NlibWidget.prototype.updateModel = function(resp){
	this.debug(NLIB_INFO,"updateModel");
	//this.debug(NLIB_JSON,"updateModel","DATA BEFORE: "+this.debugJSON(this.data));
	var data = $.parseJSON(resp); // store model
	$.extend(this.data,data)
	//this.debug(NLIB_JSON,"updateModel","DATA AFTER: "+this.debugJSON(this.data));
}

NlibWidget.prototype.updateView = function(resp){
	this.debug(NLIB_INFO,"updateView");
	if(!this.template)
		this.loadTemplate();
	this.render();
}

NlibWidget.prototype.refreshData = function(data){
  this.debug(NLIB_INFO,'refreshData');
  this.data = undefined;
  this.data = data;
  // TODO: Overload this is subclass, for example like this: (or whatever
	//       follow-up your refresh request is...)
	//this.sendRequest(this.data.followUp[0],this.encodeRequest(new Object));
}



NlibWidget.prototype.loadTemplate = function(data){
	this.debug(NLIB_INFO,"loadTemplate");
	this.template = new NlibTemplate(this.id_widget,this.src);
}

NlibWidget.prototype.render = function(data){
	this.debug(NLIB_INFO,"render");
		if(data)
		{
			this.template.render(this.id_target,data);
		}
		else
		{
			this.template.render(this.id_target,this.data);
		}
	}

NlibWidget.prototype.go = function(obj){
	this.debug(NLIB_INFO,"go");
	// TODO: Overwrite this method in subclass
	this.loadTemplate();
	this.render();
}

NlibWidget.prototype.requestOkHandler = function(obj){
	this.debug(NLIB_INFO,"requestOkHandler");
	// TODO: Overwrite this method in subclass
}

NlibWidget.prototype.requestFailedHandler = function(obj){
	this.debug(NLIB_INFO,"requestFailedHandler");
	// TODO: Overwrite this method in subclass
}

NlibWidget.prototype.updateView = function(obj){
	this.debug(NLIB_INFO,"updateView");
	if(!this.template)
		this.loadTemplate();
	this.render();
}

NlibWidget.prototype.rebind = function(obj){
	this.debug(NLIB_INFO,"rebind");
}



NlibWidget.prototype.parseResponse = function(resp){
	this.debug(NLIB_INFO,"parseResponse");
	
	//this.updateModel(resp);	
/*
	this._json = $.parseJSON(resp);
	this._status = this._json.status;
	this._running = this._json.running;

	this.debug(NLIB_DEBUG,"parseResponse","Y Response status: "+this._json.status);
	if(this._json.status == "OK")
		this.requestOkHandler();
	else
		this.requestFailedHandler();
*/

  this._json;
  this._status;
  this._running;
  this._message;

    this.debug(NLIB_DEBUG,"parseResponse","HIER");

  try{
    this._json = $.parseJSON(resp);
    this._status = this._json.status;
    this._running = this._json.running;
    this._message = this._json.message;
  } catch (e){
    this._status = "ERROR";
    this._running = 0;
    this.debug(NLIB_DEBUG,"parseResponse","Y Response status: "+this._status+", "+e.message);
    location.href = "%%%MAKEFILE_CGIBIN%%%";
  }

  if(this._status == "OK")
    this.requestOkHandler(this);
  else
    this.requestFailedHandler(this);


};

NlibWidget.prototype.sendRequest = function(req,val){
	this.debug(NLIB_INFO,"sendRequest");
	if(this.xmlhttp){
		var params = null;
		if(val)
			params = req+"="+val;
			//this.xmlhttp.open("GET", "%%%MAKEFILE_CGIBIN%%%?"+req+"="+val, true);
		this.xmlhttp.open("POST", "%%%MAKEFILE_CGIBIN%%%", true);
		this.xmlhttp.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
//		else
//			this.xmlhttp.open("GET", "%%%MAKEFILE_CGIBIN%%%?", true);
		this.xmlhttp.send(params);
	}
}

	

/*
 * A Template class
 */
NlibTemplate.prototype = new NlibObject;
NlibTemplate.prototype.constructor = NlibTemplate;
function NlibTemplate(id_template,src){
	NlibObject.call(this);
	this.src = src || "";
	this.tpl = "";
	this.id_template = id_template;
	this.loadTpl(this);
}

NlibTemplate.prototype.loadTpl = function(obj){
	this.debug(NLIB_INFO,"loadTpl");
		$.ajax({url:this.src,async:false}).done(function(tpldata){obj.tpl = tpldata;});
		//alert("tpl: "+obj.tpl);
	}

NlibTemplate.prototype.render = function(id_target,data){
	this.debug(NLIB_INFO,"render");
	//this.debug(NLIB_DEBUG,"render",this.tpl+", "+this.id_template+", "+data.widgetName);
		if(data){
			$(id_target).html(this.tpl);
			Tempo.prepare(this.id_template).render(data);	
		}
		else
		{
			this.debug(NLIB_ERROR,"render","data is undefined, required template not loaded, maybe the wrong main template?");
		}
}




NlibWidgetLogin.prototype = new NlibWidget;
NlibWidgetLogin.prototype.constructor = NlibWidgetLogin;
function NlibWidgetLogin(app,data){
	this.debug(NLIB_INFO,"new");
	this.widgetName = "Login";
	NlibWidget.call(this,this.widgetName,app,data);
}

NlibWidgetLogin.prototype.go = function(obj){
	this.debug(NLIB_INFO,"go");
	
		this.loadTemplate();
		this.render();
		
		// TODO: Overwrite this method in subclass
		$("#blogin").click(function(){
			var data = new Object();
			data.username = $("#username").val();
			data.password = $("#password").val();
			this.debug(NLIB_DEBUG,"go",obj.data.followUp[0]);
			obj.sendRequest(obj.data.followUp[0],obj.encodeRequest(data))
		});
		$("#blogout").click(function(){
			var data = new Object();
			data.username = $("#username").val();
			data.password = $("#password").val();
			obj.sendRequest(obj.data.followUp[0],obj.encodeRequest(data))
		});

	}

NlibWidgetLogin.prototype.requestOkHandler = function(obj){
	this.debug(NLIB_INFO,"requestOkHandler");
	// TODO: Overwrite this method in subclass
}

NlibWidgetLogin.prototype.requestFailedHandler = function(obj){
	this.debug(NLIB_INFO,"requestFailedHandler");
}


NlibApplication.prototype = new NlibObject;
NlibApplication.prototype.constructor = NlibApplication;
function NlibApplication(){
	this.debug(NLIB_INFO,"new");
	this._followUp = $.parseJSON($(json).html());
	this._json; //= $.parseJSON($(json).html());
	this._status = "ERROR";
	this._running = 0;
	this._widgets = new Array();


	NlibObject.call(this);
}

NlibApplication.prototype.updateAllWidgets = function(){
	this.debug(NLIB_INFO,"updateAllWidgets");
	var data = new Object();
	this.sendRequest(this._followUp.followUp[0],this.encodeRequest(data));
}


NlibApplication.prototype.run = function(){
	this.debug(NLIB_INFO,"run");
	//this.initWidgets();
	this.updateAllWidgets();
	
}

NlibApplication.prototype.isRunning = function(){
	this.debug(NLIB_INFO,"isRunning");
	return this._running;
}


NlibApplication.prototype.requestOkHandler = function(obj){
	this.debug(NLIB_INFO,"requestOkHandler");

	for(var i = 0; i<this._json.widgets.length;i++){
		var curWidget = this._json.widgets[i];
		this.debug(NLIB_DEBUG,"requestOkHandler","curWidget: "+ curWidget.widgetName);
		// dispatch widget by name
		// NEU:
		if(curWidget.widgetName == "Login"){
			if(this._widgets.Login){
				this.debug(NLIB_DEBUG,"requestOkHandler","curWidget: "+ curWidget.widgetName+" is already defined.");
			}
			else{
				this.debug(NLIB_DEBUG,"requestOkHandler","curWidget: "+ curWidget.widgetName+" is undefined. Creating new instance.");
				this._widgets.Login = new NlibWidgetLogin(this,curWidget);
				this._widgets.Login.go(this._widgets.Login);
			}
		}

		/* // ALT:
		if(curWidget.widgetName == "Login"){
			var w = new NlibWidgetLogin(this,curWidget);
			this._widgets.push(w);
			w.go(w);
		}
		*/
	}
}

NlibApplication.prototype.requestFailedHandler = function(obj){
	this.debug(NLIB_INFO,"requestFailedHandler");
}


NlibApplication.prototype.parseResponse = function(resp){
	this.debug(NLIB_INFO,"parseResponse");
	//this.followUp = this._json.followUp;
	this._json = $.parseJSON(resp);
	this._status = this._json.status;
	this._running = this._json.running;

	this.debug(NLIB_DEBUG,"parseResponse","Response status: "+this._json.status);
	if(this._json.status == "OK")
		this.requestOkHandler(this);
	else
		this.requestFailedHandler(this);
};




